motionplan/cBiRRT.go

```diff
@@ -461,59 +458,69 @@ func (mp *cBiRRTMotionPlanner) smoothPath(
 	schan := make(chan node, 1)
 	defer close(schan)
 
-	for iter := 0; iter < toIter && len(inputSteps) > 4; iter++ {
-		select {
-		case <-ctx.Done():
-			mp.logger.Debug("CBiRRT timed out during smoothing, returning best path")
-			return inputSteps
-		default:
-		}
-		// Pick two random non-adjacent indices, excepting the ends
-
-		j := 2 + mp.randseed.Intn(len(inputSteps)-3)
-
-		i := mp.randseed.Intn(j) + 1
-
-		ok, hitCorners := smoothable(inputSteps, i, j, mp.corners)
-		if !ok {
-			continue
-		}
-
-		shortcutGoal := make(map[node]node)
-
-		iSol := inputSteps[i]
-		jSol := inputSteps[j]
-		shortcutGoal[jSol] = nil
-
-		// extend backwards for convenience later. Should work equally well in both directions
-		mp.constrainedExtend(ctx, mp.randseed, shortcutGoal, jSol, iSol, schan)
-		reached := <-schan
-
-		// Note this could technically replace paths with "longer" paths i.e. with more waypoints.
-		// However, smoothed paths are invariably more intuitive and smooth, and lend themselves to future shortening,
-		// so we allow elongation here.
-		dist := mp.planOpts.DistanceFunc(&Segment{StartConfiguration: inputSteps[i].Q(), EndConfiguration: reached.Q()})
-		if dist < mp.algOpts.JointSolveDist && len(reached.Q()) < j-i {
-			mp.corners[iSol] = true
-			mp.corners[jSol] = true
-			for _, hitCorner := range hitCorners {
-				mp.corners[hitCorner] = false
+	for numCornersToPass := 2; numCornersToPass > 0; numCornersToPass-- {
+		for iter := 0; iter < toIter/2 && len(inputSteps) > 4; iter++ {
+			select {
+			case <-ctx.Done():
+				return inputSteps
+			default:
 			}
-			newInputSteps := append([]node{}, inputSteps[:i]...)
-			for reached != nil {
-				newInputSteps = append(newInputSteps, reached)
-				reached = shortcutGoal[reached]
+			// get start node of first edge. Cannot be either the last or second-to-last node.
+			// Intn will return an int in the half-open interval half-open interval [0,n)
+			i := mp.randseed.Intn(len(inputSteps) - 2)
+			j := i + 1
+			cornersPassed := 0
+			hitCorners := []node{}
+			for (cornersPassed != numCornersToPass || !inputSteps[j].Corner()) && j < len(inputSteps)-1 {
```

Comment:
I can't really figure out what kind of data structure would make this work. for constant time lookup of which bucket a node is in, buckets would have to be a Set. However then we won't know where in the segment the node is since the Set is unordered and node `i` isn't guaranteed to be a corner. if we want to guarantee `i` is a corner then i think the data structure is more straightforward but then finding `i` in the first place is not

Reply:
We could here take advantage of the fact that in Go, index-based lookups on a slice are *also* constant time. If we built, for example, a list of bools of equal length to the path, checking whether the 63rd node is a corner would be `boolList[63]`.

Reply:
Also, this looks like this will stop extending `j` at the very next node after hitting its final corner, despite there being N nodes to the following corner. It's not clear to me this is what we want; it feels arbitrary.

Reply:
`j` stops when it is a corner that occurs after we have passed `numCornersToPass` corners. so if `numCornersToPass` is 2, `j` will be the third corner to occur after `i`. I think I implemented this correctly unless there's a bug I'm not catching.

Reply:
You're correct; I misread the code.

I'm not sure that ensuring that j is always a corner (or corner-adjacent) is the best choice. Could we try something similar to the RRT* smoother and select for `j` the node which is either 25, 50, or 75% of the way to the third corner?

Reply:
resolved discussion about dividing path into buckets offline. determined that it requires a lot of changes to other parts of the cBiRRT algorithm and so won't do for this PR.